### ISSUE TITLE
False positive with E1136 unsubscriptable-object

### ISSUE THREAD

**Issue by sam-s** [PyCQA/pylint#1498]
### Steps to reproduce
1. Create file `z.py`
```
a = None
while True:
    if a is None or a["1"] == 0:
        a = {"1":1}
    else:
        break
print "Done"
```
2. run `pylint z.py`

### Current behavior

```
E:  3,20: Value 'a' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

no errors, no warnings

### pylint --version output

```
pylint 1.7.1, 
astroid 1.5.2
Python 2.7.13 (default, Dec 18 2016, 07:03:39) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

**Comment by erjiang** [comment]
This has hit us too. We are using a variable that is initialized to None, but then gets assigned to dict in a loop.

I think something like this will trigger it too:

```
points = [{"lat": 39, "lon": -92}, {"lat": 39.1, "lon": -92.1}]

furthest_west = None
for p in points:
    if furthest_west is None:
        furthest_west = p
    else:
        if p['lon'] < furthest_west['lon']:
            furthest_west = p
```

not sure how you would solve this, seems like the inferencer would have to go into the if/else clauses to see what might get assigned to the variable to figure out that it's Union(None, dict).

**Comment by hughes** [comment]
Even checking `isinstance(thing, list)` immediately prior to using `thing` will trigger this.

**Comment by PCManticore** [comment]
Yes @hughes that's a problem that's still crippling pylint even today (but hopefully with the new number of committers we have, we might tackle it at some point). I'm referring here to the fact that pylint does not quite grasp control flow, that is, if you use `isinstance`, pylint will happily infer all potential values that a variable could have, disregarding the branch hints.

**Comment by Ch3ck** [comment]
I had this problem today opening a config file.

**Comment by kelvinau** [comment]
+1

**Comment by sdmunozsierra** [comment]
I have this false positive too. I reference a variable instanced as none in the object class but  then changed it in another function.
Code example below; incomplete for brevity.
``` python
class Response():
    def __init__(self):
        self.raw_response = None

    def filter_raw_response(self):
        if not self.raw_response:
            return None
        num_datapoints = len(self.raw_response['Datapoints'])   # Error shows here
        if num_datapoints == 0:
            return None

class SpecificResponse(Response):

    def craft_response(self)
        # Variable instanced here:
        self.raw_response = self.client.get_metric_statistics()
```

**Comment by Charlie818** [comment]
+1 

I got this too when I tried to iterate over candidates, a matrix like variable. It reports item as NoneType and raise E1136 error. But, actually, it is a list.
```
candidates = [item for item in val_dataset if item[-1] == i]
candidates = random.sample(candidates,100)
for item in candidates:
    img_path = os.path.join(data_dir,item[0])
    out_path = os.path.join(output_folder,str(i))
    os.system('cp %s %s'%(img_path,out_path))
```

**Comment by LucaMoschella** [comment]
False positive even if the custom class implements the `__getitem__` method.

**Comment by PCManticore** [comment]
@melikyuksel Unless someone proposes a patch, it's unlikely.

**Comment by nilselde** [comment]
> +1
> 
> I got this too when I tried to iterate over candidates, a matrix like variable. It reports item as NoneType and raise E1136 error. But, actually, it is a list.
> 
> ```
> candidates = [item for item in val_dataset if item[-1] == i]
> candidates = random.sample(candidates,100)
> for item in candidates:
>     img_path = os.path.join(data_dir,item[0])
>     out_path = os.path.join(output_folder,str(i))
>     os.system('cp %s %s'%(img_path,out_path))
> ```

I had the same issue when using `random.sample`

**Comment by sgornick** [comment]
Would this actually be a false positive?  Seems to me it is following the intended behavior.

> Used when there is a reference to an object that initialized as a non-subscriptable object

http://pylint-messages.wikidot.com/messages:e1136

**Comment by ravinderkhatri** [comment]
The solution there is suggestion to used [] to initialize an empty list instead of using None to remove the error but doing that cause an unwanted/undesirable behavior in python. 
See here
https://nikos7am.com/posts/mutable-default-arguments/

**Comment by lxbarbosa** [comment]
@ravinderkhatri 

Thanks. [] fixed it!

**Comment by andy-maier** [comment]
There is also issue #2063 that seems to describe the same error.

**Comment by andy-maier** [comment]
I also get this with the built-in `Exception.args` and believe it is the same error:

```
class HTTPError(Exception):

    def __init__(self, status, reason):
        super(HTTPError, self).__init__(status, reason)

    @property
    def status(self):
        return self.args[0]  # pylint reports: E1136: Value 'self.args' is unsubscriptable

    @property
    def reason(self):
        return self.args[1]  # same
```

**Comment by alexyarochkin** [comment]
Just make your list Python3 compatible:
i.e.:
my_list = [1, 2, 3, 4]
for _my in list(my_list):
    pass

**Comment by jakeleventhal** [comment]
I'm also seeing this with __getitem__ implemented

**Comment by technillogue** [comment]
a similar thing also occurs with [bidict](https://bidict.readthedocs.io/en/master/)

**Comment by jab** [comment]
> a similar thing also occurs with bidict

Bidict author here. Just noticed this spurious error from pylint when using Python 3.9 and found my way to this issue. 

```
(.venv39) jab@tsmbp ~/s/bidict (dev)> pre-commit run -a
...
pylint...................................................................Failed
- hook id: pylint
- exit code: 2

************* Module bidict._typing
bidict/_typing.py:17:17: E1136: Value '_t.Union' is unsubscriptable (unsubscriptable-object)
bidict/_typing.py:20:6: E1136: Value '_t.Union' is unsubscriptable (unsubscriptable-object)
bidict/_typing.py:32:6: E1136: Value '_t.Union' is unsubscriptable (unsubscriptable-object)
bidict/_typing.py:33:6: E1136: Value '_t.Union' is unsubscriptable (unsubscriptable-object)
```

In all these cases, `_t.Union` refers to `typing.Union`, which is in fact subscriptable. See for example [this code](https://github.com/jab/bidict/blob/94b2676/bidict/_typing.py#L17) (picking the first line flagged above, `bidict/_typing.py:17`, as an example).

This result is given by pylint 2.6.0 (as can be seen [here](https://github.com/jab/bidict/blob/94b267637ec48e99c40430dac8f5b30e8ef2ebd5/.pre-commit-config.yaml#L48)), the latest stable release.

The spurious error does not occur when using the same version of pylint under Python 3.6, 3.7, or 3.8.

Hope this helps.

**Comment by hippo91** [comment]
@technillogue / @jab thanks for your report/investigation.
The problem you mention is a bit different from the original one (i.e not due to a lack of control flow) but linked to `python 3.9`.
It seems to be the same issue as #3882.

**Comment by ricardoquesada** [comment]
Happens when using hinting:

```
def my_func(rgb: tuple[int, int, int]):
   ....
```

same error: `E1136: Value 'tuple' is unsubscriptable (unsubscriptable-object)`

**Comment by ricardoquesada** [comment]
@kbakk good point.

I was just copy&pasting code from here: https://docs.python.org/3/library/typing.html

**Comment by hippo91** [comment]
@kbakk @ricardoquesada thanks for your report.
In fact, starting with `python3.9`, `tuple` is indeed subscriptable.
With master branch of `pylint` and `astroid`, when linting the following code:
```
def my_func(rgb: tuple[int]) -> None:
    rgb = None
    return
```
with `python3.7`:
```
pylint --disable=all --enable=unsubscriptable-object bug_pylint_1498.py 
************* Module bug_pylint_1498
bug_pylint_1498.py:1:17: E1136: Value 'tuple' is unsubscriptable (unsubscriptable-object)

---------------------------------------------------------------------
Your code has been rated at -6.67/10 (previous run: 10.00/10, -16.67)
```
and with `python3.9`:
```
pylint --disable=all --enable=unsubscriptable-object bug_pylint_1498.py 

---------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: -6.67/10, +16.67)
```
I'm not closing this issue, because original issue is not the same as the one we are dealing with here.

**Comment by BenLatham** [comment]
This issue also appears where the variable in question is a class atribute, as in the following minimal example:

```python
class MyClass:
    my_cashed_value = None

    def get_message(self):
        result = MyClass.my_cashed_value
        if result is not None:
            return result["message"]
        else:
            message = "my cached message"
            MyClass.my_cashed_value = {"message": message}
            return message
```

**Comment by hippo91** [comment]
@BenLatham thank for your report.
In your case, `pylint` is understanding that `my_cashed_value` maybe `None` because of its initialization.
If you change `my_cashed_value = None` for `my_cashed_value = dict()` then the issue disappears.

**Comment by andrea-cassioli-maersk** [comment]
Similar issue here using python3.8. In this example
```
"""
whatever
"""
from typing import Optional, Dict


def func(i) -> Optional[Dict]:
    """
    func
    """
    return {i: i} if i else None


x = func(1)
print(x[1])

x = func(0)
print(x[0])
```

I get 

```
09:20 $ pylint test.py 
************* Module test
test.py:18:6: E1136: Value 'x' is unsubscriptable (unsubscriptable-object)

-------------------------------------------------------------------
Your code has been rated at 2.86/10 (previous run: -1.43/10, +4.29)
```

on the second call to `func`. `pylint` is quite smart, but in more complex examples it seems to give up and simply report a variable that might be `None` to be `unsubscriptable`. Is that the case?

**Comment by Pierre-Sassoulas** [comment]
Yes, @andrea-cassioli-maersk astroid do not have real control flow (no restriction on the inference after ``isinstance(x, dict)`` for example) but can handle simple cases like the one you gave.

**Comment by belm0** [comment]
I'm still not understanding why pylint can infer this case correctly:

```python
def func(i) -> Optional[Tuple]:
    return (i, i) if i else None

print(func(1)[0])  # OK
print(func(0)[0])  # unsubscriptable-object
```

but not this:
```python
foo: Optional[Tuple] = None
print(foo and foo[0])  # unsubscriptable-object  (??)
```

**Comment by DanielNoord** [comment]
I think we can infer that in the first call of the first example `bool(i)` evaluates to `True` and thus `(i, i)` will be returned. Similarly for the second line of that example.

For the second example we only infer that `foo` is `None`. We (currently) do nothing with the typing added to `foo`. Based on the inference of `None` the message is emitted.

**Comment by belm0** [comment]
I see.

It would be a small quality of life improvement, but even without using the type annotation, pylint could certainly infer that `foo` can't be `None` (the initialized value), acknowledge that it doesn't know the type, and refrain from omitting the error.

**Comment by DanielNoord** [comment]
Perhaps I don't understand your examples but in both examples the `unsubscriptable-object` seems correct? In the first `bool(0)` will be `False` and thus `func` returns `None` whereas in the second example `foo` is initialised as `None` and isn't reassigned before the following line.

**Comment by belm0** [comment]
the expression `foo and foo[0]` implies that foo can't possibly be indexed if it's `None`

but to clarify the example:

```python
foo: Optional[Tuple] = None
# ... code that assigns a tuple to foo
print(foo and foo[0])  # unsubscriptable-object  (??)
```

**Comment by DanielNoord** [comment]
Ah my bad, that's indeed quite obvious and a pattern we use ourselves. Should have examined the example better. 

However, it is not quite that easy to add such control flow recognition to `pylint`. In fact, right now (I think) we do nothing of that type of control flow. Personally I have started to add some flow recognition for `if ... elif`, which is easier since we can check the parent node and check whether the current node is the child of an `If`.
The control flow needed for this example is something like checking whether we are in `nodes.Comparison`, whether the object being subscripted is in any of the preceding nodes of the `Comparison`, whether any of those nodes imply that the object is subscriptable, for example, if object is an `int` it would not be subscriptable but `isinstance(foo, Tuple)` would. Without any existing systems adding such control flow recognition is quite a daunting task and probably one of the reasons why this issue is still open and why we have so many open control flow issues.

**Comment by belm0** [comment]
Another short-term idea:  without fulling supporting type hints, if pylint could at least see that the var is declared as a union including at least one indexible type, it should not emit `unsubscriptable-object` due to the high chance of false positive.

**Comment by pstahlhofen** [comment]
+1
In my case, the object that is said to be unsubscriptable even though it actually is subscriptable is returned by a function from a library, so I cannot do any type-hinting or such to get rid of the wrong error message.

**Comment by matkoniecz** [comment]
following code also triggers it

maybe disable it by default as very likely to produce false positives?

```
    all_entries = []
    for key in ["surface", "surface:note"]:
        for entry in dubious_surfaces_requiring_notes.surface_values_that_trigger_note_creation(key):
            value = entry[0]
            comment = entry[1]
            all_entries.append((key, value, comment))
    for data in random.sample(all_entries, 6):
        print(data)
        key = data[0]
        value = data[1]
        comment = data[2]
        scan_for_tag(key, value, comment)
```

**Comment by optimaltayfun** [comment]
I am getting this false positive as well. Below is the code triggering this:


```python
from sqlalchemy.orm import Mapped
class UserDB(Base):
    """
    User database model.
    """
    __tablename__ = "users"
    user_id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True, autoincrement=True, nullable=False)
```
The documentation from sqlalchemy describes how to use Mapped in the following link:
https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html

But this is causing false positives with pylint:
`app/user_auth/database.py:49:13: E1136: Value 'Mapped' is unsubscriptable (unsubscriptable-object)`

**Comment by nkalpakis21** [comment]
i see this has been open for awhile. is there trouble getting a dev to fix this issue? what is blocking this for so long

**Comment by krispharper** [comment]
Like @optimaltayfun and several others, I'm getting this while upgrading to SQLAlchemy 2.0. Lots of other people will hit this doing the same thing, I think.

**Comment by matkoniecz** [comment]
@nkalpakis21 

> what is blocking this for so long

because noone implemented fix for this

you can submit pull request fixing this, hire someone to do this or hope that someone else will do either

as someone who is developer on open source projects: this kind of comment is neither helpful nor encouraging, and if someone is doing it as unpaid hobby (or poorly paid one) it will sooner or later result in them dropping such project altogether

note that ability to skip unhelpful/broken rules is present in pylint (and I use it to suppress many rules I consider unhelpful), you can use it here

**Comment by nkalpakis21** [comment]
> [@nkalpakis21](https://github.com/nkalpakis21)
> 
> > what is blocking this for so long
> 
> because noone implemented fix for this
> 
> you can submit pull request fixing this, hire someone to do this or hope that someone else will do either
> 
> as someone who is developer on open source projects: this kind of comment is neither helpful nor encouraging, and if someone is doing it as unpaid hobby (or poorly paid one) it will sooner or later result in them dropping such project altogether
> 
> note that ability to skip unhelpful/broken rules is present in pylint (and I use it to suppress many rules I consider unhelpful), you can use it here
> 
> (and this issue has both [Help wanted "🙏Outside help would be appreciated, good for new contributors"](https://github.com/pylint-dev/pylint/labels/Help%20wanted%20%F0%9F%99%8F) and [Needs PR "This issue is accepted, sufficiently specified and now needs an implementation"](https://github.com/pylint-dev/pylint/labels/Needs%20PR) labels)

sorry, I was genuinely trying to understand the bottleneck to the solution

**Comment by Pierre-Sassoulas** [comment]
Control flow is hard, it requires generic and intimidating changes in astroid (the underlying code representation of pylint) and it's inference engine (the infer function on each node of the astroid's abstract syntax tree) that is going to affect a lot of code, thus a variety of very strange and unexpected code too. If the code is too clever or too dynamic it's harder (I don't know if that's the case here). The volunteers maintainers of this project are already busy trying to follow the upstream rhythm of changes from python core to support the latest interpreter, and keeping up with the reviews of all proposed MR takes time too. (Tidelift pays me to fix security issues and prevent supply chain attack, which I do, but everything else is us giving away free time). Offering a bounty somewhere (issue hunt, github sponsor, etc.) or implementing the change yourself would probably speed things up.

**Comment by nhairs** [comment]
Hi @Pierre-Sassoulas,

Would this fall into your [$1000 one-time medium bug](https://github.com/sponsors/Pierre-Sassoulas?frequency=one-time) category?

For everyone else:

I'd be happy to help sponsor getting this fixed but I can't really afford to do it myself. Maybe we can band together 10-20 people who can all donate a bit to help get us there (if you'd be willing to donate $50-100 maybe react with a 🚀 to this comment).

**Comment by Pierre-Sassoulas** [comment]
This issue would requires a big new feature in astroid (make astroid understand control flow better) and would fix numerous nearly duplicate issues, so it's probably more akin to a large project. To be honest at this point I have too much on my plate to handle this even for a large bounty, but maybe @jacobtylerwalls or @DanielNoord could be interested (or someone on IssueHunt, from a country with low cost of living for whom the monetary compensation would be worth months of work).

**Comment by jacobtylerwalls** [comment]
https://github.com/pylint-dev/pylint/issues/1498#issuecomment-1888928977 is a separate issue from the original post. I'll open a new issue for it.

**Comment by jacobtylerwalls** [comment]
Actually, that SQLAlchemy issue looks like #9549, which has a workaround at https://github.com/pylint-dev/pylint/issues/9549#issuecomment-2429093464.
